Weird race condition with targets

## Describe the Bug

I'm hitting a weird race condition with the `GitInfo` target due to incremental build `Inputs` and `Outputs` short-circuiting targets causing their returned properties to _sometimes_ be empty.

Specifically, I'm trying to add a custom target that runs after the `GitInfo` target to use its properties to derive a base version using the `GitCommitDate`; however sometimes this property is empty due to being skipped based on incremental build `Inputs` and `Outputs` properties.

I've been investigating this weirdness with [MSBuild Log Viewer](https://github.com/KirillOsenkov/MSBuildStructuredLog) and other properties associated with the `GitInfo` target are also randomly affected.

## Steps to Reproduce

Add the following files to the root directory next to the solution file:

`Directory.Build.props`:
```xml
<Project>
  ...
  <Import Project="Version.targets" />
</Project>
```

`Version.targets`:
```xml
<?xml version="1.0" encoding="utf-8"?>
<Project>

    <PropertyGroup>
        <GitBranchSlugLength Condition="'$(GitBranchSlugLength)' == ''">0</GitBranchSlugLength>
        <!-- <GitCachePath>$([MSBuild]::MakeRelative($(MSBuildProjectDirectory), $(SolutionDir)))</GitCachePath> -->
        <!-- <GitCommitDateUtcVersion Condition="'$(GitCommitDateUtcVersion)' == ''">true</GitCommitDateUtcVersion> -->
        <GitInfoReportImportance Condition="'$(Configuration)' == 'Debug'">high</GitInfoReportImportance>
        <GitThisAssemblyMetadata Condition="'$(GitThisAssemblyMetadata)' == ''">true</GitThisAssemblyMetadata>
        <GitVersion>false</GitVersion>
    </PropertyGroup>

    <ItemGroup>
        <PackageReference Include="GitInfo" Version="3.3.3" PrivateAssets="all" />
    </ItemGroup>

    <Target Name="_GitBaseVersionCommitDate" Returns="$(GitBaseVersion)"
        AfterTargets="GitInfo"
        Condition="'$(GitBaseVersion)' == '' And '$(GitIgnoreCommitDateVersion)' != 'true'">

        <PropertyGroup Condition="'$(GitCommitDateUtcVersion)' == 'true'">
            <GitBaseVersion>$([System.DateTimeOffset]::Parse("$(GitCommitDate)").ToUniversalTime().ToString("yyyy.Mdd.Hmm"))</GitBaseVersion>
            <GitBaseVersionSource>GitCommitDateUtc</GitBaseVersionSource>
            <GitSemVerSource>CommitDateUtc</GitSemVerSource>
        </PropertyGroup>

        <PropertyGroup Condition="'$(GitCommitDateUtcVersion)' != 'true'">
            <GitBaseVersion>$([System.DateTimeOffset]::Parse("$(GitCommitDate)").ToString("yyyy.Mdd.Hmm"))</GitBaseVersion>
            <GitBaseVersionSource>GitCommitDate</GitBaseVersionSource>
            <GitSemVerSource>CommitDate</GitSemVerSource>
        </PropertyGroup>

        <Exec Command="$(GitExe) rev-list --count --first-parent $(GitDefaultBranch)..HEAD"
            EchoOff="true"
            StandardErrorImportance="low"
            StandardOutputImportance="low"
            ConsoleToMSBuild="true"
            WorkingDirectory="$(GitRoot)"
            ContinueOnError="true"
            StdOutEncoding="utf-8">
            <Output TaskParameter="ConsoleOutput" PropertyName="GitCommits" />
            <Output TaskParameter="ExitCode" PropertyName="MSBuildLastExitCode" />
        </Exec>

        <PropertyGroup Condition="'$(MSBuildLastExitCode)' != '0'">
            <GitCommits>0</GitCommits>
        </PropertyGroup>

    </Target>

    <Target Name="_GitSetVersion"
        AfterTargets="GitVersion"
        Condition="'$(GitVersion)' == 'false'">

        <PropertyGroup>
            <_BranchSlug>$(GitBranch.ToLowerInvariant())</_BranchSlug>
            <_BranchSlug Condition="$(GitBranchSlugLength) > 0 And $(_BranchSlug.Length) > $(GitBranchSlugLength)">$(_BranchSlug.Substring(0, $(GitBranchSlugLength)))</_BranchSlug>
            <_BranchSlug>$([System.Text.RegularExpressions.Regex]::Replace($(_BranchSlug), "[^0-9a-z-]", "-").Trim('-'))</_BranchSlug>
            <Version>$(GitBaseVersionMajor).0.0.0</Version>
            <FileVersion>$(GitBaseVersionMajor).$(GitBaseVersionMinor).$(GitBaseVersionPatch).0</FileVersion>
            <InformationalVersion Condition="'$(GitBranch)' == '$(GitDefaultBranch)'">$(GitBaseVersionMajor).$(GitBaseVersionMinor).$(GitBaseVersionPatch)+$(_BranchSlug).$(GitSha)</InformationalVersion>
            <InformationalVersion Condition="'$(GitBranch)' != '$(GitDefaultBranch)'">$(GitBaseVersionMajor).$(GitBaseVersionMinor).$(GitBaseVersionPatch)-$(_BranchSlug).$(GitCommits)</InformationalVersion>
            <PackageVersion Condition="'$(GitBranch)' == '$(GitDefaultBranch)'">$(GitBaseVersionMajor).$(GitBaseVersionMinor).$(GitBaseVersionPatch)</PackageVersion>
            <PackageVersion Condition="'$(GitBranch)' != '$(GitDefaultBranch)'">$(GitBaseVersionMajor).$(GitBaseVersionMinor).$(GitBaseVersionPatch)-$(_BranchSlug).$(GitCommits)</PackageVersion>
        </PropertyGroup>

    </Target>

</Project>
```

In my scenario I have a solution with only 4 projects but they all want to generate the same version - which highlights another issue that the git info cache default directory should probably be common to all projects; either in the `.git` directory (preferred) like [GitVersion](https://github.com/GitTools/GitVersion) or next to the solution file. Is there any reason this is not the default?

Anyway, if I repeatedly _rebuild_ the solution it eventually fails, usually within the first 10 attempts, with the following:
```
error MSB4184: The expression "[System.DateTimeOffset]::Parse('')" cannot be evaluated. String '' was not recognized as a valid DateTime.
```

## Expected Behavior

Always succeed rather than randomly fail.

## Version Info

`3.3.3`

## Additional Info

MSBuild Log Viewer screenshot (redacted):

![image](https://github.com/devlooped/GitInfo/assets/750121/1486f295-370b-4784-8ed2-62dabe8aeaae)

Shows that after the `GitInfo` target was run, several of its properties, including `GitCommitDate` where I expand its target `_GitCommitDate`, were not populated due to the incremental build skipping them:
```
Skipping target ... because all output files are up-to-date with respect to the input files.
```

Looking around this repo's source it seems to me that the targets that populate the `GitInfo` properties do not need to leverage the incremental build properties (`Inputs` and `Outputs`) since they are also short circuited by their conditions when the cache is hit and the values can be read from the previous build.

Am I doing something wrong or missing something?

PS. I'm happy to submit a pull request if we can understand the issue and decide how best to fix it. Along the same lines, would you accept a pull request to add this "git commit date" based versioning strategy once I get it working robustly? 🙏

See https://github.com/gitfool/GitInfo.Dungeon for a repro. I [injected some sleeps](https://github.com/gitfool/GitInfo.Dungeon/blob/bcfdad60b162eb2a67b31377cfc7c902b818e609/Version.targets#L78-L80) after `_GitReadCache` for a slightly more deterministic build that flushes out the race condition between `_GitReadCache` and `_GitWriteCache`, when using root level cache files.

It doesn't want to repro within 10 test builds on github but it always fails within 10 runs on my local machine:
```
❯ .\test-build.ps1

**************************************************
Build #1
VERBOSE: Performing the operation "Remove File" on target "D:\Devel\gitfool\GitInfo.Dungeon\.GitHead.cache".
VERBOSE: Performing the operation "Remove File" on target "D:\Devel\gitfool\GitInfo.Dungeon\.GitInfo.cache".
VERBOSE: Performing the operation "Remove File" on target "D:\Devel\gitfool\GitInfo.Dungeon\.GitIsDirty.cache".
Shutting down MSBuild server...
Shutting down VB/C# compiler server...
VB/C# compiler server shut down successfully.
MSBuild server shut down successfully.
Tool 'cake.tool' (version '3.1.0') was restored. Available commands: dotnet-cake

Restore was successful.
Analyzing build script...
Processing build script...
Installing tools...
Tool dotnet-reportgenerator-globaltool is already installed, with required version.
Tool GitVersion.Tool is already installed, with required version.
Installing addins...
Running cached build script...
Executing: "D:/Devel/gitfool/GitInfo.Dungeon/tools/dotnet-gitversion.exe" -output json -verbosity Verbose

========================================
Info
========================================
Executing task: Info
Version 0.1.0+4.Branch.main.Sha.bcfdad60b162eb2a67b31377cfc7c902b818e609
Finished executing task: Info
Completed in 00:00:00.0308242

========================================
BuildSolutions
========================================
Executing task: BuildSolutions
Cleaning directory D:/Devel/gitfool/GitInfo.Dungeon/Cli/bin
Cleaning directory D:/Devel/gitfool/GitInfo.Dungeon/Library/bin
Cleaning directory D:/Devel/gitfool/GitInfo.Dungeon/Tests/bin
Cleaning directory D:/Devel/gitfool/GitInfo.Dungeon/Cli/obj
Cleaning directory D:/Devel/gitfool/GitInfo.Dungeon/Library/obj
Cleaning directory D:/Devel/gitfool/GitInfo.Dungeon/Tests/obj
Executing: "C:/Program Files/dotnet/dotnet.exe" build "D:/Devel/gitfool/GitInfo.Dungeon/GitInfo.Dungeon.sln" --configuration Release /property:Version=0.0.0.0 /property:FileVersion=0.1.0.0 /property:InformationalVersion=0.1.0+4.Branch.main.Sha.bcfdad60b162eb2a67b31377cfc7c902b818e609 /property:PackageVersion=0.1.0 /property:EmbedAllSources=true /property:RestoreLockedMode=false /property:ContinuousIntegrationBuild=false
MSBuild version 17.8.0+6cdef4241 for .NET
  Determining projects to restore...
  Restored D:\Devel\gitfool\GitInfo.Dungeon\Cli\Cli.csproj (in 244 ms).
  Restored D:\Devel\gitfool\GitInfo.Dungeon\Library\Library.csproj (in 244 ms).
  Restored D:\Devel\gitfool\GitInfo.Dungeon\Tests\Tests.csproj (in 449 ms).
  Sleeping 1ms...
  Sleeping 1ms...
  Sleeping 1000ms...
D:\Devel\gitfool\GitInfo.Dungeon\Version.targets(15,13): error MSB4184: The expression "[System.DateTimeOffset]::Parse('')" cannot be evaluated. String '' was not recognized as a valid DateTime. [D:\Devel\gitfool\GitInfo.Dungeon\Tests\Tests.csproj]
  Library -> D:\Devel\gitfool\GitInfo.Dungeon\Library\bin\Release\net7.0\Library.dll
  Cli -> D:\Devel\gitfool\GitInfo.Dungeon\Cli\bin\Release\net7.0\Cli.dll

Build FAILED.

D:\Devel\gitfool\GitInfo.Dungeon\Version.targets(15,13): error MSB4184: The expression "[System.DateTimeOffset]::Parse('')" cannot be evaluated. String '' was not recognized as a valid DateTime. [D:\Devel\gitfool\GitInfo.Dungeon\Tests\Tests.csproj]
    0 Warning(s)
    1 Error(s)

Time Elapsed 00:00:05.08
An error occurred when executing task 'BuildSolutions'.
Completed in 00:00:05.3824110
Error: .NET CLI: Process returned an error (exit code 1).
```

--- a/src/GitInfo/build/GitInfo.targets
+++ b/src/GitInfo/build/GitInfo.targets
@@ -384,12 +384,12 @@
     <Delete Files="$(_GitInfoFile)" />
   </Target>
 
-  <Target Name="_GitReadCache" Condition="Exists('$(_GitInfoFile)') And '$(SkipReadGitCache)' != 'true' ">
-    <PropertyGroup>
-      <_GitCachedInfo>$([System.IO.File]::ReadAllText('$(_GitInfoFile)'))</_GitCachedInfo>
-    </PropertyGroup>
+  <Target Name="_GitReadCache" Condition="Exists('$(_GitInfoFile)') And '$(SkipReadGitCache)' != 'true'">
+    <ReadLinesFromFile File="$(_GitInfoFile)" ContinueOnError="true">
+      <Output TaskParameter="Lines" ItemName="_GitCachedInfo" />
+    </ReadLinesFromFile>
 
-    <CreateItem Include="GitInfo" AdditionalMetadata="$(_GitCachedInfo.Split(';'))">
+    <CreateItem Include="GitInfo" AdditionalMetadata="@(_GitCachedInfo -> TrimEnd(';'))">
       <Output TaskParameter="Include" ItemName="GitInfo" />
     </CreateItem>
 

--- a/src/GitInfo/build/GitInfo.targets
+++ b/src/GitInfo/build/GitInfo.targets
@@ -150,8 +150,11 @@
     <!-- Cache file used to avoid running all git commands. Only GitRoot will be retrieved to determine the path of this cache file. -->
     <_GitInfoFile>$(GitCachePath)GitInfo.cache</_GitInfoFile>
 
-    <!-- Stores value of GitIsDirty variable from last execution, file is changed when the value changes betwee runs. -->
+    <!-- Stores value of GitIsDirty variable from last execution. File is changed when the value changes between runs. -->
     <_GitIsDirtyFile>$(GitCachePath)GitIsDirty.cache</_GitIsDirtyFile>
+
+    <!-- Stores value of _GitHeadPath variable from last execution. Added as an input for the fast up-to-date check VS performs. -->
+    <_GitHeadFile>$(GitCachePath)GitHead.cache</_GitHeadFile>
   </PropertyGroup>
 
   <Target Name="GitInfoReport" DependsOnTargets="GitInfo;GitVersion">
@@ -978,25 +981,23 @@
 
   </Target>
 
-  <!-- This combination of prop/item/target fix incremental builds after HEAD changes. -->
-  <PropertyGroup>
-    <_GitCacheFullPath>$([System.IO.Path]::Combine($(MSBuildProjectDirectory), $(GitCachePath)))</_GitCacheFullPath>
-    <!-- We read the path to the HEAD file, written in a previously run _GitWriteHeadPath target -->
-    <_GitHeadPath Condition="Exists('$(_GitCacheFullPath)GitHead.cache')">$([System.IO.File]::ReadAllText('$(_GitCacheFullPath)GitHead.cache').Trim())</_GitHeadPath>
-  </PropertyGroup>
+  <!-- This combination of target/item/target fixes incremental builds after HEAD changes. -->
+  <Target Name="_GitReadHeadPath" AfterTargets="GitVersion">
+    <ReadLinesFromFile File="$(_GitHeadFile)" ContinueOnError="true">
+      <Output TaskParameter="Lines" ItemName="_GitHeadPath" />
+    </ReadLinesFromFile>
+  </Target>
 
   <ItemGroup>
     <!-- And if the HEAD path exists, we add it as an input for the fast up-to-date check VS performs -->
-    <UpToDateCheckInput Condition="'$(_GitHeadPath)' != '' and Exists('$(_GitHeadPath)')" Include="$(_GitHeadPath)" />
+    <UpToDateCheckInput Condition="'@(_GitHeadPath)' != '' and Exists('@(_GitHeadPath)')" Include="@(_GitHeadPath)" />
   </ItemGroup>
 
   <!-- NOTE: we write the path to HEAD after first calculating a version, since we can be anywhere in
        the solution structure. This file will only be written if there's a change in the git root,
        which should never happen. -->
   <Target Name="_GitWriteHeadPath" AfterTargets="GitVersion">
-    <WriteLinesToFile Lines="$(GitRoot).git\HEAD"
-						  File="$(GitCachePath)GitHead.cache"
-						  Overwrite="true" WriteOnlyWhenDifferent="true" />
+    <WriteLinesToFile File="$(_GitHeadFile)" Lines="$(GitRoot).git\HEAD" Overwrite="true" WriteOnlyWhenDifferent="true" ContinueOnError="true" />
   </Target>
 
   <!--

--- a/src/GitInfo/build/GitInfo.targets
+++ b/src/GitInfo/build/GitInfo.targets
@@ -12,17 +12,17 @@
 
 	Customization:
 
-  $(GitVersion): set to 'false' to avoid setting Version and PackageVersion 
-                 to a default version with format: 
+  $(GitVersion): set to 'false' to avoid setting Version and PackageVersion
+                 to a default version with format:
                  $(GitSemVerMajor).$(GitSemVerMinor).$(GitSemVerPatch)$(GitSemVerDashLabel)+$(GitBranch).$(GitCommit)
 
 	$(GitThisAssembly): set to 'false' to prevent assembly
 						metadata and constants generation.
 
-  $(GitThisAssemblyMetadata): set to 'false' to prevent assembly 
-                              metadata generation only. Defaults 
-                              to 'false'. If 'true', it will also 
-                              provide assembly metadata attributes 
+  $(GitThisAssemblyMetadata): set to 'false' to prevent assembly
+                              metadata generation only. Defaults
+                              to 'false'. If 'true', it will also
+                              provide assembly metadata attributes
                               for each of the populated values.
 
 	$(ThisAssemblyNamespace): allows overriding the namespace
@@ -70,7 +70,7 @@
 					 in a previous build in a GitInfo.cache for
 					 performance reasons.
 					 Defaults to empty value (no ignoring).
-					 
+
   $(GitCachePath): where to cache the determined Git information
 				   gives the chance to use a shared location
 				   for different projects. this can improve
@@ -78,19 +78,19 @@
 				   has to end with a path seperator
 				   Defaults to empty value ('$(IntermediateOutputPath)').
 
-	$(GitTagRegex): Regular expression used with git describe to filter the tags 
+	$(GitTagRegex): Regular expression used with git describe to filter the tags
 					to consider for base version lookup.
 					Defaults to * (all)
-           
+
 	$(GitBaseVersionRegex): Regular expression used to match and validate valid base versions
-							in branch, tag or file sources. By default, matches any string that 
+							in branch, tag or file sources. By default, matches any string that
 							*ends* in a valid SemVer2 string.
 							Defaults to 'v?(?<MAJOR>\d+)\.(?<MINOR>\d+)\.(?<PATCH>\d+)(?:\-(?<LABEL>[\dA-Za-z\-\.]+))?$|^(?<LABEL>[\dA-Za-z\-\.]+)\-v?(?<MAJOR>\d+)\.(?<MINOR>\d+)\.(?<PATCH>\d+)$'
-	
+
 	$(GitCommitDateFormat): The value passed as the format option when trying to retrieve
 							the git commit date
 							Defaults to %%cI (windows) or %cI (non windows)
-           
+
 	==============================================================
 	-->
 
@@ -414,9 +414,7 @@
   </Target>
 
   <Target Name="_GitRepositoryUrl" Returns="$(GitRepositoryUrl)" DependsOnTargets="_GitRoot"
-			Condition="'$(GitRoot)' != '' and '$(GitRepositoryUrl)' == ''"
-			Inputs="@(_GitInput)"
-			Outputs="$(_GitInfoFile)">
+			Condition="'$(GitRoot)' != '' and '$(GitRepositoryUrl)' == ''">
 
     <Exec Command="$(GitExe) config --get remote.$(GitRemote).url"
 			  Condition="'$(GitRepositoryUrl)' == ''"
@@ -444,8 +442,6 @@
 
   <Target Name="_GitCommitDate" Returns="$(GitCommitDate)"
 			DependsOnTargets="_GitRoot"
-			Inputs="@(_GitInput)"
-			Outputs="$(_GitInfoFile)"
 			Condition="'$(GitRoot)' != '' And '$(GitCommitDate)' == ''">
 
     <Exec Command='$(GitExe) -c log.showSignature=false show --format=$(GitCommitDateFormat) -s'
@@ -467,10 +463,8 @@
   </Target>
 
   <Target Name="_GitBranch" Returns="$(GitBranch)"
-					  DependsOnTargets="_GitInputs"
-			Condition="'$(GitBranch)' == ''"
-			Inputs="@(_GitInput)"
-			Outputs="$(_GitInfoFile)">
+			DependsOnTargets="_GitInputs"
+			Condition="'$(GitBranch)' == ''">
 
     <Exec Command='$(GitExe) rev-parse --abbrev-ref HEAD'
 			Condition="'$(GitRoot)' != ''"
@@ -487,7 +481,7 @@
 
     <!-- CI systems may checkout the specific commit, rather than the branch, so we need to fallback -->
     <Exec Command='$(GitExe) name-rev --name-only $(GitNameRevOptions) HEAD'
-			  Condition=" '$(GitBranch)' == '' "
+			  Condition="'$(GitRoot)' != '' and '$(GitBranch)' == ''"
 			  EchoOff='true'
 			  StandardErrorImportance="low"
 			  StandardOutputImportance="low"
@@ -506,8 +500,6 @@
 
   <Target Name="_GitCommit" Returns="$(GitCommit)"
 			DependsOnTargets="_GitRoot"
-			Inputs="@(_GitInput)"
-			Outputs="$(_GitInfoFile)"
 			Condition="'$(GitRoot)' != '' And '$(GitCommit)' == ''">
 
     <Exec Command='$(GitExe) -c log.showSignature=false log --format=format:$(_ShortShaFormat) -n 1'
@@ -578,7 +570,7 @@
           <Version>$(GitSemVerMajor).$(GitSemVerMinor).$(GitSemVerPatch)$(GitSemVerDashLabel)+$([System.Text.RegularExpressions.Regex]::Replace($(GitBranch), "[^0-9A-Za-z-]", "-")).$(GitCommit)</Version>
           <PackageVersion>$(Version)</PackageVersion>
       </PropertyGroup>
-  </Target>  
+  </Target>
 
   <Target Name="_GitBaseVersionFile" Returns="$(GitBaseVersion)"
 			Inputs="@(_GitInput)"
@@ -998,8 +990,8 @@
     <UpToDateCheckInput Condition="'$(_GitHeadPath)' != '' and Exists('$(_GitHeadPath)')" Include="$(_GitHeadPath)" />
   </ItemGroup>
 
-  <!-- NOTE: we write the path to HEAD after first calculating a version, since we can be anywhere in 
-       the solution structure. This file will only be written if there's a change in the git root, 
+  <!-- NOTE: we write the path to HEAD after first calculating a version, since we can be anywhere in
+       the solution structure. This file will only be written if there's a change in the git root,
        which should never happen. -->
   <Target Name="_GitWriteHeadPath" AfterTargets="GitVersion">
     <WriteLinesToFile Lines="$(GitRoot).git\HEAD"

--- a/src/GitInfo/build/GitInfo.targets
+++ b/src/GitInfo/build/GitInfo.targets
@@ -354,7 +354,7 @@
       <GitIsDirty>0</GitIsDirty>
     </PropertyGroup>
 
-    <WriteLinesToFile File="$(_GitIsDirtyFile)" Lines="$(GitIsDirty)" Overwrite="true" WriteOnlyWhenDifferent="true"/>
+    <WriteLinesToFile File="$(_GitIsDirtyFile)" Lines="$(GitIsDirty)" Overwrite="true" WriteOnlyWhenDifferent="true" ContinueOnError="true" />
   </Target>
 
   <Target Name="_GitInputs" DependsOnTargets="_GitRoot" Returns="@(_GitInput)">

